File: Rate Limiter/middleware/rateLimiter.js
const redis = require('redis');
const fs = require('fs');
const path = require('path');

const client = redis.createClient({
    url: 'redis://localhost:6379'
});

// Connecting to Redis
(async () => {
    await client.connect();
})();

const BUCKET_SIZE = 210;
const REFILL_RATE = 10;   // Tokens per second
const MAX_RETRIES = 3;    // Maximum retries for race conditions

// Load Lua script
const luaScript = fs.readFileSync(path.join(__dirname, 'rateLimiter.lua'), 'utf8');

const rateLimiter = async (req, res, next) => {
    const key = `ratelimit:${req.ip}`;
    const now = Date.now();
    let retries = 0;

    const attemptRateLimit = async () => {
        try {
            const result = await client.eval(
                luaScript,
                {
                    keys: [key],
                    arguments: [String(now), String(REFILL_RATE)]
                }
            );

            const [remaining, retryAfter] = result;

            // Set standard rate limit headers
            res.set({
                'X-RateLimit-Limit': BUCKET_SIZE,
                'X-RateLimit-Remaining': Math.max(0, remaining),
                'X-RateLimit-Reset': now + (retryAfter * 1000)
            });

            if (remaining >= 0) {
                return next();
            } else if (remaining === -2 && retries < MAX_RETRIES) {
                // Handle race condition
                retries++;
                return new Promise(resolve => setTimeout(resolve, 100))
                    .then(() => attemptRateLimit());
            } else {
                // Rate limit exceeded
                res.set('X-RateLimit-Retry-After', Math.ceil(retryAfter));
                return res.status(429).json({
                    error: 'Too Many Requests',
                    retryAfter: Math.ceil(retryAfter)
                });
            }
        } catch (error) {
            console.error('Rate limiter error:', error);
            return res.status(500).json({ error: 'Internal Server Error' });
        }
    };

    return attemptRateLimit();
};

module.exports = rateLimiter;

File: Rate Limiter/middleware/rateLimiter.lua
-- rateLimiter.lua
local key = KEYS[1]
local now = tonumber(ARGV[1])
local refillRate = tonumber(ARGV[2])
local bucketSize = 210  -- As per assignment requirement

-- Get current bucket state
local bucket = redis.call('hmget', key, 'tokens', 'lastRefill', 'lockVersion')
local tokens = tonumber(bucket[1])
local lastRefill = tonumber(bucket[2])
local lockVersion = tonumber(bucket[3])

-- Initialize if not exists
if not tokens then
    tokens = bucketSize
    lastRefill = now
    lockVersion = 0
    redis.call('hmset', key, 'tokens', tokens, 'lastRefill', lastRefill, 'lockVersion', lockVersion)
end

-- Calculate token refill
local elapsedTime = (now - lastRefill) / 1000
local tokensToAdd = math.floor(elapsedTime * refillRate)
local newTokens = math.min(bucketSize, tokens + tokensToAdd)

-- Attempt atomic operation with optimistic locking
if newTokens > 0 then
    -- Increment lock version to handle race condition
    local newLockVersion = lockVersion + 1
    
    -- Check if the state hasn't changed (race condition check)
    local current = redis.call('hget', key, 'lockVersion')
    if tonumber(current) == lockVersion then
        -- Update bucket state atomically
        redis.call('hmset', key, 
            'tokens', newTokens - 1,
            'lastRefill', now,
            'lockVersion', newLockVersion
        )
        return {newTokens - 1, 0}  -- {remaining tokens, retry after}
    else
        return {-2, 0.1}  -- Race condition detected, retry after 100ms
    end
else
    -- Calculate retry after time
    local timeToNextToken = (1 - (newTokens % 1)) / refillRate
    return {-1, timeToNextToken}  -- Rate limit exceeded
end